File: app/src/main/java/igrek/todotree/service/tree/TreeSelectionManager.java
package igrek.todotree.service.tree;


import java.util.TreeSet;

public class TreeSelectionManager {
	
	private TreeSet<Integer> selectedPositions = null;
	
	public TreeSet<Integer> getSelectedItems() {
		return selectedPositions;
	}
	
	public TreeSet<Integer> getSelectedItemsNotNull() {
		return selectedPositions != null ? selectedPositions : new TreeSet<Integer>();
	}
	
	public int getSelectedItemsCount() {
		if (selectedPositions == null)
			return 0;
		return selectedPositions.size();
	}
	
	public boolean isAnythingSelected() {
		return selectedPositions != null && selectedPositions.size() > 0;
	}
	
	public void startSelectionMode() {
		selectedPositions = new TreeSet<>();
	}
	
	public void cancelSelectionMode() {
		selectedPositions = null;
	}
	
	public void setItemSelected(int position, boolean selectedState) {
		if (!isAnythingSelected()) {
			startSelectionMode();
		}
		if (selectedState) {
			if (!isItemSelected(position)) {
				selectedPositions.add(position);
			}
		} else {
			if (isItemSelected(position)) {
				selectedPositions.remove(Integer.valueOf(position));
				if (selectedPositions.isEmpty()) {
					selectedPositions = null;
				}
			}
		}
	}
	
	private boolean isItemSelected(int position) {
		for (Integer pos : selectedPositions) {
			if (pos == position) {
				return true;
			}
		}
		return false;
	}
	
	public void toggleItemSelected(int position) {
		setItemSelected(position, !isItemSelected(position));
	}
	
}

File: app/src/main/java/igrek/todotree/service/tree/TreeSelectionManager.kt
package igrek.todotree.service.tree

import java.util.TreeSet

class TreeSelectionManager {

    var selectedItems: TreeSet<Int>? = null
        private set
    val selectedItemsNotNull: TreeSet<Int>
        get() = if (selectedItems != null) selectedItems!! else TreeSet()
    val selectedItemsCount: Int
        get() = if (selectedItems == null) 0 else selectedItems!!.size
    val isAnythingSelected: Boolean
        get() = selectedItems != null && selectedItems!!.size > 0

    fun startSelectionMode() {
        selectedItems = TreeSet()
    }

    fun cancelSelectionMode() {
        selectedItems = null
    }

    fun setItemSelected(position: Int, selectedState: Boolean): Boolean {
        // Return true if all elements has to be refreshed
        var refreshAll = false
        if (!isAnythingSelected) {
            startSelectionMode()
            refreshAll = true
        }
        if (selectedState) {
            if (!isItemSelected(position)) {
                selectedItems?.add(position)
            }
        } else {
            if (isItemSelected(position)) {
                selectedItems?.remove(Integer.valueOf(position))
                if (selectedItems.isNullOrEmpty()) {
                    selectedItems = null
                    refreshAll = true
                }
            }
        }
        return refreshAll
    }

    private fun isItemSelected(position: Int): Boolean {
        for (pos in selectedItems!!) {
            if (pos == position) {
                return true
            }
        }
        return false
    }

    fun toggleItemSelected(position: Int): Boolean {
        return setItemSelected(position, !isItemSelected(position))
    }
}